### ISSUE TITLE
Training command/weights

### ISSUE THREAD

**Issue by sshleifer** [jungokasai/deep-shallow#1]
Do you have a training command handy or pretrained weights?
These models seem very good and fast!

**Comment by jungokasai** [comment]
Thank you so much for your interest in our work! I updated the [readme file](https://github.com/jungokasai/deep-shallow/blob/master/README.md) with commands and links to pretrained models and datasets.

**Comment by sshleifer** [comment]
Thanks a ton! Do you have the big teacher models handy?

**Comment by jungokasai** [comment]
Yes, check `transformer.wmt16.en-de` in the fairseq [readme](https://github.com/pytorch/fairseq/blob/master/examples/translation/README.md#pre-trained-models)! Note though, when people do knowledge distillation, they train a teacher model on the training data and run it on the same training data. So teacher training is done for each training data.

**Comment by sshleifer** [comment]
Cool, that makes sense. Thx!

**Comment by sshleifer** [comment]
Do you have other wmt dataset links by any chance? (seems like you ran en-fr and en-zh from the paper).
I am trying to assemble them for the huggingface nlp/datasets package.
Will credit you appropriately!

**Comment by jungokasai** [comment]
That sounds like a great idea! Please cite the paper if you think that is appropriate. And let me know if I can help you with this further! 
- [WMT14 EN-FR](https://drive.google.com/uc?id=1F77xiMa-oW_PPPCox-2OmQzBn1mmJVPG)
- [WMT16 EN-RO](https://drive.google.com/uc?id=1comPmyGQPDVM-HiqqN-uXD_7hIP-Qsf7)
- [WMT17 EN-ZH](https://drive.google.com/uc?id=1niWAlufwojjcCOV2frKQekyMoHs1QqrE)

**Comment by sshleifer** [comment]
Yes will cite, you are so helpful!

My only further ask is more checkpoints. We can put them in our the huggingface modelhub under allennlp (which I infer is the relevant organization?), and copy your README.md as a model card if that's of interest? 

We're actually not quite ready on our end, but will likely be soon when the currently in progress fairseq converter is done https://github.com/huggingface/transformers/issues/5419.

**Benefits:**
- you'd get a webpage like [this](https://huggingface.co/Helsinki-NLP/opus-mt-en-de?text=My+name+is+Sarah+and+I+live+in+London) (The markdown on the bottom can just be your original README + a link to the original REPO + citation info, whatever you want)
- Some extra exposure, cites and usage from us tweeting + all our users using your model(s).

**Comment by sshleifer** [comment]
Out of curiosity, why did you guys use `transformer.wmt16.en-de` as the teacher instead of `transformer.wmt19.en-de`?
cc @stas00

**Comment by jungokasai** [comment]
Good question! I used WMT16 training data for comparisons with prior works. But I've recently trained Enc6-Dec6 models for WMT19 de-en with fairseq, so I'll share that here. 
Note that this is WMT19 `de-en`, and test data for WMT 19 `de-en` and `en-de` are [different to avoid _translationese_ effects](https://www.aclweb.org/anthology/W19-5301/)! I believe this wasn't the case with WMT16/14.
* [dataset](https://drive.google.com/uc?id=1j6z9fYdlUyOYsh7KJoumRlr1yHczxR5T): [moses](https://github.com/moses-smt/mosesdecoder/tree/master/scripts/tokenizer) tokenized followed by bpe with [fastbpe](https://github.com/glample/fastBPE)) 
* [fairseq transformer base](https://drive.google.com/uc?id=1yT7ZjqfvUYOBXvMjeY8uGRHQFWoSo8Q5)
* [fairseq transformer big](https://drive.google.com/uc?id=15gAzHeRUCs-QV8vHeTReMPEh1j8excNE)

It'd be fantastic if we can convert fairseq models to the HuggingFace interface, something like the [Helsinki-NLP pipeline](https://huggingface.co/Helsinki-NLP).

**Comment by stas00** [comment]
Awesome! Thank you, @jungokasai 

I will work on that. We are waiting for the [initial model](https://github.com/huggingface/transformers/pull/6940) to be merged, and then will be able to add your models.

Please put this data with links on some page in this repo, so that we could link to it as the source.

Let us know what would you like to be written on each of the model cards - we will now have 5 different models of yours in 2 groups.

p.s. the 3 models that are described on the front page are also moses+fastbpe, correct? At least the saved args say so. Just wanted to make sure it is so.

**Comment by stas00** [comment]
also let's pick a short unique string to identify all your models by, e.g. "kasai"? e.g.:

```
fsmt-kasai-wmt16-en-de-dist-12-1
fsmt-kasai-wmt16-en-de-dist-6-1
fsmt-kasai-wmt16-en-de-12-1
fsmt-kasai-wmt19-de-en-6-6-base
fsmt-kasai-wmt19-de-en-6-6-big
```
`fsmt` is the internal arch name, the rest is up to you. This is what I tentatively picked up. Feel free to make a different suggestion.

The API will be:

```
mname = "allen_nlp/fsmt-kasai-wmt16-en-de-dist-12-1"
tokenizer = FSMTTokenizer.from_pretrained(mname)
model = FSMTForConditionalGeneration.from_pretrained(mname)

text = "Maschinelles Lernen ist großartig, oder?"
encoded = tokenizer.encode(text, return_tensors='pt')
outputs = model.generate(encoded, num_beams=5, num_return_sequences=5, early_stopping=True)
for i, output in enumerate(outputs):
    decoded = tokenizer.decode(output, skip_special_tokens=True)
```

For example for fairseq transformer we currently have (name is a not final):
```
fsmt-wmt19-en-de
fsmt-wmt19-de-en
fsmt-wmt19-en-ru
fsmt-wmt19-ru-en
```
but probably should inject an extra identifier as well to disambiguate, perhaps `fsmt-orig` or something?

```
fsmt-orig-wmt19-en-de
fsmt-orig-wmt19-de-en
fsmt-orig-wmt19-en-ru
fsmt-orig-wmt19-ru-en
```


@sshleifer, what do you think?

I know that they will eventually reside in different orgs `facebook/...`, `allen_nlp/...`, but still it'd be easier to have the dir-names consistent and distinct.

**Comment by stas00** [comment]
OK, ported wmt19 big/base, got these bleu scores with wmt19 test data:
base:
```
{'bleu': 37.6868, 'n_obs': 2000, 'runtime': 102, 'seconds_per_sample': 0.051}
```
big:
```
{'bleu': 39.1698, 'n_obs': 2000, 'runtime': 190, 'seconds_per_sample': 0.095}
```

For comparison fairseq transformers wmt19 de-en reports 42.3. And our ported model of the same currently gets 41.18. (all on the same test data)

I think our model may need a bit more work, but that's a good start. We are also behind since we have no re-ranking in place, just the best score in the beam. Also note that fairseq transformers uses a 4 model ensemble, which we currently can't support either.

What do you get on your side, @jungokasai?

**Comment by sshleifer** [comment]
+ Keep naming as short as possible, no fsmt, no kasai: `allen_nlp/wmt16-en-de-12-1`.
+ For each size, I see no value in porting second best checkpoints.
+ Cite their deep-shallow paper in the model card, and just copy relevant portions of this REPO's README.

**Comment by jungokasai** [comment]
> OK, ported wmt19 big/base, got these bleu scores with wmt19 test data:
> base:
> 
> ```
> {'bleu': 37.6868, 'n_obs': 2000, 'runtime': 102, 'seconds_per_sample': 0.051}
> ```
> 
> big:
> 
> ```
> {'bleu': 39.1698, 'n_obs': 2000, 'runtime': 190, 'seconds_per_sample': 0.095}
> ```
> 
> For comparison fairseq transformers wmt19 de-en reports 42.3. And our ported model of the same currently gets 41.18. (all on the same test data)
> 
> I think our model may need a bit more work, but that's a good start. We are also behind since we have no re-ranking in place, just the best score in the beam. Also note that fairseq transformers uses a 4 model ensemble, which we currently can't support either.
> 
> What do you get on your side, @jungokasai?

Without any compound splitting, I got `base=38.34` and `big=40.62` with `checkpoint_last3_avg.pt`.

**Comment by stas00** [comment]
Thank you for sharing the eval results, @jungokasai 

Indeed, `checkpoint_last3_avg.pt` scores better - I have re-converted based on it and now I get slightly better scores:
```
base=38.02
big=39.32
```
So we are pretty close!

Do you get these scores with re-ranking or just generate? I'm trying to figure out where we are missing that last bit of precision.

I can't tell whether your `generate` code is slightly different from fairseq's master as yours was forked some months ago - but my guess is that it must be the same.

And I will switch to that better checkpoint then. `checkpoint_best.pt` proved to be not the best for the 2nd time ;)

**Comment by jungokasai** [comment]
Yes, I also believe my fork uses the same decoding method. It is just beam search and no reranking happening. 
```bash
python generate.py <path_to_data_dir> --path <path_to_model> --remove-bpe --max-tokens 8000 --beam 5 --lenpen 0.6
```
Everything else is default. For example, `lenpen=0.6`. That can be the difference?

**Comment by stas00** [comment]
> For example, lenpen=0.6. That can be the difference?

Yes, I got a 0.3-0.5 boost using `lenpen=0.6`, instead of the default 1.0. So we are at:

```
base=38.37
big=39.89
```
with the converted model. Yay! Thank you for that suggestion, @jungokasai 

So we get the exact same score on `base` as you, but are short by 0.73 on `big` :( which is odd - do you think there could have been something else that isn't the same? Perhaps the online `big` model is different from your local one you have just tested with, or we aren't testing the same checkpoint file?

**Comment by stas00** [comment]
We are also trying to find that information for fairseq wmt19 transformer - the paper mentions a tuned up length penalty but I can't find what that was. Perhaps, if you have some insights/guesses that would be of a great help.

**Comment by stas00** [comment]
I tested `lenpen=0.6` on your wmt16 models and also getting an improvement - should we use this setting for both of your sets as the default?
https://github.com/huggingface/transformers/issues/7049#issuecomment-691620767

p.s. strangely I still get a better score using `checkpoint_best.pt` for `wmt16-en-de-dist-12-1` rather than `checkpoint_top5_average.pt`, but the latter scores better for the other 2 models. Are those online files all up-to-date?

**Comment by jungokasai** [comment]
Hmmn, I still got `40.62` with the online version. Are you measuring BLEU after detokenization? I'm measuring tokenized BLEU as that's the output I get with fairseq `generate.py`.

**Comment by stas00** [comment]
Thank you for validating that, @jungokasai 

We have a slightly different process, so yes, we are scoring on the the de-tokenized, user-ready final output.

I will have to step through with debugger and see where we diverge. 

Since your models are not on `torch.hub`, what would be the equivalent of this code - so that I could do th e same with your models? I run this side-by-side with our code and compare things step by step.

```
    import torch
    text =  "Maschinelles Lernen ist großartig, oder?",
    mname = f"transformer.wmt19.de-en"
    checkpoint_file = 'model4.pt'
    model = torch.hub.load('pytorch/fairseq', mname, checkpoint_file=checkpoint_file)
    encoded = model.encode(text)
    outputs = model.generate(encoded, beam=num_beams)
    decoded = model.decode(outputs[0]['tokens'])
```
Thanks.

**Comment by jungokasai** [comment]
Right, detokenized BLEU is typically 0.5+ points lower than moses tokenized BLEU (e.g., see Table 3 in [Ott et al. 2018](https://arxiv.org/pdf/1806.00187.pdf)). So I'm not too worried about the discrepancy we are seeing. An equivalent script for our custom model is the following:
```python
from fairseq.models.transformer import TransformerModel
model_dir = <PATH_TO_MY_MODEL_DIR>
en2de = TransformerModel.from_pretrained(model_dir, checkpoint_file='checkpoint_top5_average.pt',
    data_name_or_path=model_dir, bpe='subword_nmt', bpe_codes=model_dir+'bpecodes', tokenizer='moses')
text = 'Machine learning is great, isn't it?'
en2de.translate(text)
```
where set `bpe='fastbpe'` for the wmt19 models.

**Comment by stas00** [comment]
> detokenized BLEU is typically 0.5+ points lower than moses tokenized BLEU

Oh, that's interesting! I will have to study why this is so. One would think that the testing should be done on the final output. Thank you for this insight, @jungokasai.

> An equivalent script for our custom model is the following:

Oh, OK, that's simple - for some reason I was doing it the hard way and failing. Thanks a lot!

**Comment by jungokasai** [comment]
Yes, I also didn't realize it was this easy to do generation on the fly using a custom model on fairseq! 
One more thing I realized is that you can batch input by making a list of text. 
```python
text = ["Machine learning is great, isn't it?", "Everyone should learn machine learning."]
en2de.translate(text)
```
would also work! And regarding evaluation, yes the community is moving toward sacrebleu after detokenization now. Unfortunately there's still some legacy sticking around in what people report in papers, including mine, for comparisons with prior works.

**Comment by stas00** [comment]
We are almost there with porting of your 5 models - just need to write the model cards.

For wmt16 3 models we have https://github.com/jungokasai/deep-shallow/, so I have all the info we need.

In need your input for the 2 wmt19 models of yours:

* Where should we link to for the 2 wmt19 models of yours (as you shared the download link in a comment of this issue). Perhaps it'd be useful to make a dedicated page for it? it's totally up to you - or we can leave it w/o a link too.
* how do we cite the  2 wmt19 models of yours? is there a paper in works, or perhaps a page we could link to?

Thank you!

**Comment by jungokasai** [comment]
This is great news! For those two WMT19 models, please refer to the same deep-shallow [paper](https://arxiv.org/abs/2006.10369) and [repository](https://github.com/jungokasai/deep-shallow/)!

**Comment by stas00** [comment]
Fantastic. I was now able to complete the cards. I will ping you once things are online.

Thanks again!

**Comment by stas00** [comment]
OK, they all are online now: https://huggingface.co/models?filter=allenai&tag=fsmt

Thank you for your support, @jungokasai!

**Comment by zmf0507** [comment]
Hi @stas00 , can you tell me how can I re-produce the results on newstest2016 test set using model.generate() function ? 27.1 mentioned on the model link https://huggingface.co/allenai/wmt16-en-de-dist-6-1

  beam_output = model.generate(
        input_ids, 
        num_beams=num_beams, 
       early_stopping=False,
        length_penalty = 1
    ).to(device)

I have tried changing length penalty , early stopping and max_length but the best I could get is 20 sacreBleu for max_len = 50. The hugging-face model page only tells about using beam size 5 , implying that all other params have to be kept default. I have tried that as well and it only gave me 14 scareBleu. 
Please let me know the exact config to get the reported bleu score. Thanks

**Comment by stas00** [comment]
Please see: https://github.com/huggingface/transformers/blob/master/scripts/fsmt/eval-allenai-wmt16.sh
This is the code I run to get the results posted on that page.

The other eval scripts for the other models are here: https://github.com/huggingface/transformers/tree/master/scripts/fsmt
